File: examples/ListVenues/ListVenues/ViewController.m
//
//  ViewController.m
//  ShowFullscreenMap
//
//  Created by Juan Kruger on 31/01/18.
//  Copyright © 2018 LocusLabs. All rights reserved.
//

#import "ViewController.h"
#import "MapController.h"

@interface ViewController ()

@property (nonatomic, strong) NSArray           *venues;
@property (nonatomic, strong) LLVenueDatabase   *venueDatabase;

@end

@implementation ViewController {
    
}

#pragma mark Lifecycle

- (void)viewDidLoad {
    
    [super viewDidLoad];
    
    self.title = @"Venues";
    
    // Initialize the LocusLabs SDK with the accountId provided by LocusLabs
    [LLLocusLabs setup].accountId = @"A11F4Y6SZRXH4X";
    
    // Load the venue database
    self.venueDatabase = [LLVenueDatabase venueDatabase];
    self.venueDatabase.delegate = self;
    
    [self.venueDatabase listVenues];
}

#pragma mark Delegates - LLVenueDatabase

- (void)venueDatabase:(LLVenueDatabase *)venueDatabase venueLoadFailed:(NSString *)venueId code:(LLDownloaderError)errorCode message:(NSString *)message {

    // Handle failures here
}

- (void)venueDatabase:(LLVenueDatabase *)venueDatabase venueList:(NSArray *)venueList {
    
    self.venues = [venueList sortedArrayUsingDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:@"name" ascending:YES]]];
    [self.tableView reloadData];
}

#pragma mark Delegates - TableView

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    return [self.venues count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {

    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];

    if (!cell) {
        
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
    }
    
    LLVenueInfo *venueInfo = self.venues[indexPath.row];
    cell.textLabel.text = [NSString stringWithFormat:@"%@ (%@)", venueInfo.name, [venueInfo.airportCode uppercaseString]];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    LLVenueInfo *venueInfo = self.venues[indexPath.row];
    
    MapController *mapController = [[MapController alloc] init];
    mapController.venueID = venueInfo.venueId;
    [self.navigationController pushViewController:mapController animated:YES];
}

@end

File: examples-swift/ListVenuesSwift/ListVenuesSwift/ViewController.swift
//
//  ViewController.swift
//  ShowFullScreenMapSwift
//
//  Created by Juan Kruger on 31/01/18.
//  Copyright © 2018 LocusLabs. All rights reserved.
//

import UIKit

class ViewController: UITableViewController, LLVenueDatabaseDelegate  {

    // Vars
    var venueDatabase:      LLVenueDatabase!
    
    var venues =             [LLVenueInfo]()
    
    // MARK: Lifecycle
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        title = "Venues"
        
        // Initialize the LocusLabs SDK with the accountId provided by LocusLabs
        LLLocusLabs.setup().accountId = "A11F4Y6SZRXH4X"
        
        // Get an instance of LLVenueDatabase, register as its delegate and load the venue LAX
        venueDatabase = LLVenueDatabase()
        venueDatabase.delegate = self
        
        // Load the venue LAX async
        venueDatabase.listVenues()
    }
    
    // MARK: Delegates - LLVenueDatabase
    func venueDatabase(_ venueDatabase: LLVenueDatabase!, venueLoadFailed venueId: String!, code errorCode: LLDownloaderError, message: String!) {
        
        // Handle failures here
    }
    
    func venueDatabase(_ venueDatabase: LLVenueDatabase!, venueList: [Any]!) {
        
        venues = venueList as! [LLVenueInfo]
        venues = venues.sorted(by: { (venue1, venue2) -> Bool in
            return venue1.name < venue2.name
        })
        
        tableView.reloadData()
    }
    
    // MARK: Delegates - TableView
    override func numberOfSections(in tableView: UITableView) -> Int {
        
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        return venues.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        var cell = tableView.dequeueReusableCell(withIdentifier: "Cell")
        
        if (cell == nil) {cell = UITableViewCell(style: .default, reuseIdentifier: "Cell")}
        
        let venueInfo = venues[indexPath.row]
        let airportCode = venueInfo.venueId.count > 3 ? String(venueInfo.venueId.prefix(3)) : venueInfo.venueId
        cell?.textLabel?.text = String.init(format: "%@ (%@)", venueInfo.name, airportCode!.uppercased())
        
        return cell!
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        tableView.deselectRow(at: indexPath, animated: true)
        
        let venueInfo = venues[indexPath.row]
        
        let mapController = MapController()
        mapController.venueID = venueInfo.venueId
        navigationController?.pushViewController(mapController, animated: true)
    }
    
}

